File: Heap/HeapEx.java
import java.util.PriorityQueue;
import java.util.Arrays;
import java.util.Collections;
import java.util.Random;
public class HeapEx {

    public static void demo(String[] args) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        // PriorityQueue<Integer> pq = new PriorityQueue<Integer>(Collections.reverseOrder());
        int[] arr = { 1, 2, 10, 8, 7, 3, 4, 6, 5, 9 };
    
        for (int i : arr) {
            pq.add(i);
        }
        System.out.println("Printing Priority Queue Heap : " + pq);
    
        System.out.print("Dequeue elements of Priority Queue ::");
        while (pq.isEmpty() == false) {
            System.out.print(" " + pq.remove());
        }
    }
    
    public static int kthSmallest(int[] arr, int size, int k) {
        Arrays.sort(arr);
        return arr[k - 1];
    }
    
    public static int kthSmallest2(int[] arr, int size, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        for (int i = 0; i < size; i++) 
            pq.add(arr[i]);
        for (int i = 0; i < k-1; i++)
            pq.remove();
        return pq.peek();
    }
    
    public static int kthSmallest3(int[] arr, int size, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>(Collections.reverseOrder());
        for (int i = 0; i < size; i++) {
            if(i < k)
                pq.add(arr[i]);
            else {
                if (pq.peek() > arr[i]) {
                    pq.add(arr[i]);
                    pq.remove();
                }
            }
        }
        return pq.peek();
    }

    public static int kthLargest(int[] arr, int size, int k) {
        int value = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>(Collections.reverseOrder());
        for (int i = 0; i < size; i++)
            pq.add(arr[i]);
        for (int i = 0; i < k; i++)
            value = pq.remove();
        return value;
    }

    public static boolean isMinHeap(int[] arr, int size) {
        int lchild, rchild;
        // last element index size - 1
        for (int parent = 0; parent < (size / 2 + 1); parent++) {
            lchild = parent * 2 + 1;
            rchild = parent * 2 + 2;
            // heap property check.
            if (((lchild < size) && (arr[parent] > arr[lchild])) || ((rchild < size) && (arr[parent] > arr[rchild])))
                return false;
        }
        return true;
    }
    
    public static boolean isMaxHeap(int[] arr, int size) {
        int lchild, rchild;
        // last element index size - 1
        for (int parent = 0; parent < (size / 2 + 1); parent++) {
            lchild = parent * 2 + 1;
            rchild = lchild + 1;
            // heap property check.
            if (((lchild < size) && (arr[parent] < arr[lchild])) || ((rchild < size) && (arr[parent] < arr[rchild])))
                return false;
        }
        return true;
    }
    
    public static void main0(String[] args) {
        int arr[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("Kth Smallest :: " + kthSmallest(arr, arr.length, 3));
        int arr2[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("Kth Smallest :: " + kthSmallest2(arr2, arr2.length, 3));
    }
    /*
    Kth Smallest :: 5
    Kth Smallest :: 5
    */
    
    public static void main1() {    
        int arr3[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("isMaxHeap :: " + isMaxHeap(arr3, arr3.length));
        int arr4[] = { 1, 2, 3, 4, 5, 6, 7, 8 };
        System.out.println("isMinHeap :: " + isMinHeap(arr4, arr4.length));
    }
    
    /*
    isMaxHeap :: true
    isMinHeap :: true     
    */
    public static int kSmallestProduct(int[] arr, int size, int k) {
        Arrays.sort(arr);
        int product = 1;
        for (int i = 0; i < k; i++)
            product *= arr[i];
        return product;
    }
    
    public static void swap(int arr[], int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    
    public static void quickSelectUtil(int[] arr, int lower, int upper, int k) {
        if (upper <= lower)
            return;
    
        int pivot = arr[lower];
        int start = lower;
        int stop = upper;
    
        while (lower < upper) {
            while (lower < upper && arr[lower] <= pivot) {
                lower++;
            }
            while (lower <= upper && arr[upper] > pivot) {
                upper--;
            }
            if (lower < upper) {
                swap(arr, upper, lower);
            }
        }
    
        swap(arr, upper, start); // upper is the pivot position
        if (k < upper)
            quickSelectUtil(arr, start, upper - 1, k); // pivot -1 is the upper for left sub array.
        if (k > upper)
            quickSelectUtil(arr, upper + 1, stop, k); // pivot + 1 is the lower for right sub array.
    }
    
    public static int kSmallestProduct3(int[] arr, int size, int k) {
        quickSelectUtil(arr, 0, size - 1, k);
        int product = 1;
        for (int i = 0; i < k; i++)
            product *= arr[i];
        return product;
    }
    
    public static int kSmallestProduct2(int[] arr, int size, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        int i = 0;
        int product = 1;
        for (i = 0; i < size; i++) {
            pq.add(arr[i]);
        }
        i = 0;
        while (i < size && i < k) {
            product *= pq.remove();
            i += 1;
        }
        return product;
    }
    
    public static int kSmallestProduct4(int[] arr, int size, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>(Collections.reverseOrder());        
        for (int i = 0; i < size; i++) {
            if(i < k)
                pq.add(arr[i]);
            else {
                if (pq.peek() > arr[i]) {
                    pq.add(arr[i]);
                    pq.remove();
                }
            }
        }    
        int product = 1;
        for (int i = 0; i < k; i++) {
            product *= pq.remove();
        }
        return product;
    }

    public static void main2() {
        int arr[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("Kth Smallest product:: " + kSmallestProduct(arr, 8, 3));
        int arr2[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("Kth Smallest product:: " + kSmallestProduct2(arr2, 8, 3));
        int arr3[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("Kth Smallest product:: " + kSmallestProduct3(arr3, 8, 3));
        int arr4[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        System.out.println("Kth Smallest product:: " + kSmallestProduct4(arr4, 8, 3));
    }

    /*
     * Kth Smallest product:: 10 
     * Kth Smallest product:: 10 
     * Kth Smallest product:: 10
     * Kth Smallest product:: 10
     */
    
    public static void printLargerHalf(int[] arr, int size) {
        Arrays.sort(arr);
        for (int i = size / 2; i < size; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
    
    public static void printLargerHalf2(int[] arr, int size) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        for (int i = 0; i < size; i++) {
            pq.add(arr[i]);
        }
    
        for (int i = 0; i < size / 2; i++)
            pq.remove();
        System.out.println(pq);
    }
    
    public static void printLargerHalf3(int[] arr, int size) {
        quickSelectUtil(arr, 0, size - 1, size / 2);
        for (int i = size / 2; i < size; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
    
    public static void main3() {
        int arr[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        printLargerHalf(arr, 8);
        int arr2[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        printLargerHalf2(arr2, 8);
        int arr3[] = { 8, 7, 6, 5, 7, 5, 2, 1 };
        printLargerHalf3(arr3, 8);
    }

    /*
     * 6 7 7 8 
     * [6, 7, 7, 8] 
     * 6 7 7 8
     */
    
    public static void sortK(int[] arr, int size, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        int i = 0;
        for (i = 0; i < k; i++) {
            pq.add(arr[i]);
        }
    
        int[] output = new int[size];
        int index = 0;
    
        for (i = k; i < size; i++) {
            output[index++] = pq.remove();
            pq.add(arr[i]);
        }
        while (pq.size() > 0)
            output[index++] = pq.remove();
    
        for (i = k; i < size; i++) {
            arr[i] = output[i];
        }
    }
    
    // Testing Code
    public static void main4() {
        int k = 3;
        int[] arr = { 1, 5, 4, 10, 50, 9 };
        int size = arr.length;
        sortK(arr, size, k);
        for (int i = 0; i < size; i++)
            System.out.print(arr[i] + " ");
    }
    /*
     * 1 5 4 9 10 50
     */
    public static void main(String[] args) {
        main2();
    }
}

File: Heap/HeapEx.js
var HeapEx = /** @class */ (function () {
    function HeapEx() {
    }
    HeapEx.demo = function (args) {
        var pq = (new java.util.PriorityQueue());
        var arr = [1, 2, 10, 8, 7, 3, 4, 6, 5, 9];
        for (var index8207 = 0; index8207 < arr.length; index8207++) {
            var i = arr[index8207];
            {
                pq.add(i);
            }
        }
        console.info("Printing Priority Queue Heap : " + pq);
        console.info("Dequeue elements of Priority Queue ::");
        while ((pq.isEmpty() === false)) {
            {
                console.info(" " + pq.remove());
            }
        }
        ;
    };
    HeapEx.KthSmallest = function (arr, size, k) {
        /* sort */ (function (l) { l.sort(); })(arr);
        return arr[k - 1];
    };
    HeapEx.KthSmallest2 = function (arr, size, k) {
        var i = 0;
        var value = 0;
        var pq = (new java.util.PriorityQueue());
        for (i = 0; i < size; i++) {
            {
                pq.add(arr[i]);
            }
            ;
        }
        while ((i < size && i < k)) {
            {
                value = pq.remove();
                i += 1;
            }
        }
        ;
        return value;
    };
    HeapEx.isMinHeap = function (arr, size) {
        var lchild;
        var rchild;
        for (var parent_1 = 0; parent_1 < ((size / 2 | 0) + 1); parent_1++) {
            {
                lchild = parent_1 * 2 + 1;
                rchild = parent_1 * 2 + 2;
                if (((lchild < size) && (arr[parent_1] > arr[lchild])) || ((rchild < size) && (arr[parent_1] > arr[rchild])))
                    return false;
            }
            ;
        }
        return true;
    };
    HeapEx.isMaxHeap = function (arr, size) {
        var lchild;
        var rchild;
        for (var parent_2 = 0; parent_2 < ((size / 2 | 0) + 1); parent_2++) {
            {
                lchild = parent_2 * 2 + 1;
                rchild = lchild + 1;
                if (((lchild < size) && (arr[parent_2] < arr[lchild])) || ((rchild < size) && (arr[parent_2] < arr[rchild])))
                    return false;
            }
            ;
        }
        return true;
    };
    HeapEx.main2 = function (args) {
        var arr = [8, 7, 6, 5, 7, 5, 2, 1];
        console.info("Kth Smallest :: " + HeapEx.KthSmallest(arr, arr.length, 3));
        var arr2 = [8, 7, 6, 5, 7, 5, 2, 1];
        console.info("Kth Smallest :: " + HeapEx.KthSmallest2(arr2, arr2.length, 3));
        var arr3 = [8, 7, 6, 5, 7, 5, 2, 1];
        console.info("isMaxHeap :: " + HeapEx.isMaxHeap(arr3, arr3.length));
        var arr4 = [8, 7, 6, 5, 7, 5, 2, 1];
        /* sort */ (function (l) { l.sort(); })(arr4);
        console.info("isMinHeap :: " + HeapEx.isMinHeap(arr4, arr4.length));
    };
    HeapEx.KSmallestProduct = function (arr, size, k) {
        /* sort */ (function (l) { l.sort(); })(arr);
        var product = 1;
        for (var i = 0; i < k; i++) {
            product *= arr[i];
        }
        return product;
    };
    HeapEx.swap = function (arr, i, j) {
        var temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    };
    HeapEx.QuickSelectUtil = function (arr, lower, upper, k) {
        if (upper <= lower)
            return;
        var pivot = arr[lower];
        var start = lower;
        var stop = upper;
        while ((lower < upper)) {
            {
                while ((lower < upper && arr[lower] <= pivot)) {
                    {
                        lower++;
                    }
                }
                ;
                while ((lower <= upper && arr[upper] > pivot)) {
                    {
                        upper--;
                    }
                }
                ;
                if (lower < upper) {
                    HeapEx.swap(arr, upper, lower);
                }
            }
        }
        ;
        HeapEx.swap(arr, upper, start);
        if (k < upper)
            HeapEx.QuickSelectUtil(arr, start, upper - 1, k);
        if (k > upper)
            HeapEx.QuickSelectUtil(arr, upper + 1, stop, k);
    };
    HeapEx.KSmallestProduct3 = function (arr, size, k) {
        HeapEx.QuickSelectUtil(arr, 0, size - 1, k);
        var product = 1;
        for (var i = 0; i < k; i++) {
            product *= arr[i];
        }
        return product;
    };
    HeapEx.KSmallestProduct2 = function (arr, size, k) {
        var pq = (new java.util.PriorityQueue());
        var i = 0;
        var product = 1;
        for (i = 0; i < size; i++) {
            {
                pq.add(arr[i]);
            }
            ;
        }
        while ((i < size && i < k)) {
            {
                product *= pq.remove();
                i += 1;
            }
        }
        ;
        return product;
    };
    HeapEx.main3 = function (args) {
        var arr = [8, 7, 6, 5, 7, 5, 2, 1];
        console.info("Kth Smallest product:: " + HeapEx.KSmallestProduct(arr, 8, 3));
        var arr2 = [8, 7, 6, 5, 7, 5, 2, 1];
        console.info("Kth Smallest product:: " + HeapEx.KSmallestProduct2(arr2, 8, 3));
        var arr3 = [8, 7, 6, 5, 7, 5, 2, 1];
        console.info("Kth Smallest product:: " + HeapEx.KSmallestProduct3(arr3, 8, 3));
    };
    HeapEx.PrintLargerHalf = function (arr, size) {
        /* sort */ (function (l) { l.sort(); })(arr);
        for (var i = (size / 2 | 0); i < size; i++) {
            console.info(arr[i]);
        }
        console.info();
    };
    HeapEx.PrintLargerHalf2 = function (arr, size) {
        var product = 1;
        var pq = (new java.util.PriorityQueue());
        for (var i = 0; i < size; i++) {
            {
                pq.add(arr[i]);
            }
            ;
        }
        for (var i = 0; i < (size / 2 | 0); i++) {
            pq.remove();
        }
        console.info(pq);
    };
    HeapEx.PrintLargerHalf3 = function (arr, size) {
        HeapEx.QuickSelectUtil(arr, 0, size - 1, (size / 2 | 0));
        for (var i = (size / 2 | 0); i < size; i++) {
            console.info(arr[i]);
        }
        console.info();
    };
    HeapEx.main4 = function (args) {
        var arr = [8, 7, 6, 5, 7, 5, 2, 1];
        HeapEx.PrintLargerHalf(arr, 8);
        var arr2 = [8, 7, 6, 5, 7, 5, 2, 1];
        HeapEx.PrintLargerHalf2(arr2, 8);
        var arr3 = [8, 7, 6, 5, 7, 5, 2, 1];
        HeapEx.PrintLargerHalf3(arr3, 8);
    };
    HeapEx.sortK = function (arr, size, k) {
        var pq = (new java.util.PriorityQueue());
        var i = 0;
        for (i = 0; i < size; i++) {
            {
                pq.add(arr[i]);
            }
            ;
        }
        var output = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(size);
        var index = 0;
        for (i = k; i < size; i++) {
            {
                output[index++] = pq.remove();
                pq.add(arr[i]);
            }
            ;
        }
        while ((pq.size() > 0)) {
            output[index++] = pq.remove();
        }
        ;
        for (i = k; i < size; i++) {
            {
                arr[i] = output[i];
            }
            ;
        }
        console.info(output);
    };
    HeapEx.main5 = function (args) {
        var k = 3;
        var arr = [1, 5, 4, 10, 50, 9];
        var size = arr.length;
        HeapEx.sortK(arr, size, k);
    };
    HeapEx.ChotaBhim = function (cups, size) {
        var time = 60;
        /* sort */ (function (l) { l.sort(); })(cups);
        var total = 0;
        var index;
        var temp;
        while ((time > 0)) {
            {
                total += cups[0];
                cups[0] = (Math.ceil(cups[0] / 2.0) | 0);
                index = 0;
                temp = cups[0];
                while ((index < size - 1 && temp < cups[index + 1])) {
                    {
                        cups[index] = cups[index + 1];
                        index += 1;
                    }
                }
                ;
                cups[index] = temp;
                time -= 1;
            }
        }
        ;
        console.info("Total %d " + total);
        return total;
    };
    HeapEx.ChotaBhim2 = function (cups, size) {
        var time = 60;
        /* sort */ (function (l) { l.sort(); })(cups);
        var total = 0;
        var i;
        var temp;
        while ((time > 0)) {
            {
                total += cups[0];
                cups[0] = (Math.ceil(cups[0] / 2.0) | 0);
                i = 0;
                while ((i < size - 1)) {
                    {
                        if (cups[i] > cups[i + 1])
                            break;
                        temp = cups[i];
                        cups[i] = cups[i + 1];
                        cups[i + 1] = temp;
                        i += 1;
                    }
                }
                ;
                time -= 1;
            }
        }
        ;
        console.info("Total : " + total);
        return total;
    };
    HeapEx.ChotaBhim3 = function (cups, size) {
        var time = 60;
        var pq = (new java.util.PriorityQueue());
        var i = 0;
        for (i = 0; i < size; i++) {
            {
                pq.add(cups[i]);
            }
            ;
        }
        var total = 0;
        var value;
        while ((time > 0)) {
            {
                value = pq.remove();
                total += value;
                value = (Math.ceil(value / 2.0) | 0);
                pq.add(value);
                time -= 1;
            }
        }
        ;
        console.info("Total : " + total);
        return total;
    };
    HeapEx.JoinRopes = function (ropes, size) {
        /* sort */ (function (l) { l.sort(); })(ropes);
        console.info(ropes);
        var total = 0;
        var value = 0;
        var temp;
        var index;
        var length = size;
        while ((length >= 2)) {
            {
                value = ropes[length - 1] + ropes[length - 2];
                total += value;
                index = length - 2;
                while ((index > 0 && ropes[index - 1] < value)) {
                    {
                        ropes[index] = ropes[index - 1];
                        index -= 1;
                    }
                }
                ;
                ropes[index] = value;
                length--;
            }
        }
        ;
        console.info("Total : " + total);
        return total;
    };
    HeapEx.JoinRopes2 = function (ropes, size) {
        var pq = (new java.util.PriorityQueue());
        var i = 0;
        for (i = 0; i < size; i++) {
            {
                pq.add(ropes[i]);
            }
            ;
        }
        var total = 0;
        var value = 0;
        while ((pq.size() > 1)) {
            {
                value = pq.remove();
                value += pq.remove();
                pq.add(value);
                total += value;
            }
        }
        ;
        console.info("Total : %d " + total);
        return total;
    };
    HeapEx.main6 = function (args) {
        var cups = [2, 1, 7, 4, 2];
        HeapEx.ChotaBhim(cups, cups.length);
        var cups2 = [2, 1, 7, 4, 2];
        HeapEx.ChotaBhim2(cups2, cups.length);
        var cups3 = [2, 1, 7, 4, 2];
        HeapEx.ChotaBhim3(cups3, cups.length);
        var ropes = [2, 1, 7, 4, 2];
        HeapEx.JoinRopes(ropes, ropes.length);
        var rope2 = [2, 1, 7, 4, 2];
        HeapEx.JoinRopes2(rope2, rope2.length);
    };
    HeapEx.kthLargestStream = function (k) {
        var pq = (new java.util.PriorityQueue());
        var size = 0;
        var data = 0;
        while ((true)) {
            {
                console.info("Enter data: ");
                if (size < k - 1)
                    pq.add(data);
                else {
                    if (size === k - 1)
                        pq.add(data);
                    else if (pq.peek() < data) {
                        pq.add(data);
                        pq.remove();
                    }
                    console.info("Kth larges element is :: " + pq.peek());
                }
                size += 1;
            }
        }
        ;
    };
    HeapEx.main = function (args) {
        HeapEx.kthLargestStream(3);
    };
    return HeapEx;
}());
HeapEx["__class"] = "HeapEx";
HeapEx.main(null);
